open-company/open-company-auth
# Subscription billing

# Do not deploy until migration plan has been finished #94 

Trello: https://trello.com/c/ZZdNpiHE

Test with:
~~https://github.com/open-company/open-company-web/pull/1054~~
~~https://github.com/open-company/open-company-mobile/pull/7~~
~~https://github.com/open-company/open-company-storage/pull/225~~
Infra dependency: ~~https://github.com/belucid/open-company-infrastructure/pull/120~~

#### Testing the happy path:

- Sign up with a new org
- [x] Was a Stripe customer created?
- [x] In the Billing panel, are you on a 14 day trial?
- Click `Add payment method`
- [x] Are you redirected to Stripe's cc entry form?
- Fill in the details. Use `4242 4242 4242 4242` as the card number, and any future date and 3-digit code for expiration and CVC.
- [x] Are you redirected back to Carrot when finished?
- [x] Do you get a `Plan Active` message upon your return?
- [x] Are you on the monthly plan? Dates/totals look okay?
- [x] In the Stripe dashboard, does your customer now have an associated cc?
- [x] In the Stripe dashboard, has the trial ended, and do you see a subscription to `Monthly`?
- Back in the Carrot billing panel, choose `Change plan`
- Switch to the `Annual` plan
- [x] Estimates look correct?
- Submit the plan change
- [x] Do you see a loading spinner while the subscription is changed?
- [x] Does the plan change reflect the new plan choice on the UI upon success?
- [x] In Stripe, was your customer _*NOT*_ billed for the plan change?
- [x] In Stripe, is the customer now subscribed to two plans, one `Monthly` ending on `Today + 30 days`, and one `Annual` set to begin on `Today + 30 days`?
- [x] Is the upcoming invoice scheduled for the right date, and for the correct total?
- Back in Carrot, invite team members to your team via email, until you reach the proration limit (12 users at the time of writing). After accepting the invites, you should have 13 total users in your org.
- [x] In Stripe, was the customer invoiced a proration amount on the current `Monthly` subscription for adding the new users?
- [x] In Stripe, does the upcoming `Annual` invoice also reflect the fact that there are 13 active users?
- Back in Carrot, remove one of the users from your organization.
- [x] In Stripe, check that a proration credit was _*NOT*_ applied
- [x] Does the scheduled `Annual` plan reflect the correct amount of users and price?

#### Testing annual plan switching:
- Sign up with a new org
- Open billing panel
- Switch to Annual
- Click Add payment method
- Close the window
- Open another window and go back to carrot
- Open billing panel
- [x] do you still see the Monthly plan selected? (you can check on stripe dashboard too if you want)
- Change to Annual again
- Click Add payment method
- Add a card and all needed data and go back to Carrot
- [x] Do you see a spinner with updating plan copy?
- [x] Are you on annual plan once that finishes?
- [x] Make sure that you see the same plan in Stripe dashboard for your customer

#### Testing trial periods:

- Create a new organization
- [x] You should be on a trial version of the monthly plan
- In Stripe, find the ID of your customer's monthly _subscription_ (*not the plan*, but the actual subscription object)
- From the auth REPL, use `(update-subscription! sub-id {"trial_end" "now"})` where `sub-id` is the subscription ID you obtained in the last step. This will end your trial period immediately.
- [x] Check that you no longer have post access in Carrot
- In the billing panel, add a payment method, and complete the workflow (just like the happy path test steps above)
- [ ] Once returned to Carrot, do you have post access again?

#### Testing team seats report:
- Sign up with a new organization via Slack (make sure it's not associated with other orgs)
- [x] Check on Stripe that a customer was created and the number of seats is 1
- Now add the bot
- Add an email domain (possibly one that you can use for testing, like carrot.io or your personal domain if you have one)
- Sign up with another member of that Slack org
- [x] Check on Stripe that the seats number has been updated (should be 2 now)
- [x] Now signup with a user using the email domain
- [x] Check on Stripe that the seats number has not been updated
- Go to your email and click on the click to verify your email address
- [x] Go to Stripe and verify that the seats number has been updated (should be 3 now)
- Now with the first user invite another Slack user
- [x] Check on Stripe that the seats number has not been updated (should be 3 still)
- [x] Now accept the invitation with the Slack user
- [x] Check on Stripe that the number of seats has been updated (should be 4 now)
- With the first user invite another user via email (a new email address, never used before)
- [x] Check on Stripe that the seats number has not been updated (should be 4 still)
- Accept the invitation
- [x] Check on Stripe that the number of seats has been updated (should be 5 now)
- Invite a user that is already on Carrot but in another team
- [x] Check on Stripe that the number of seats has been updated right away (should be 6 now)
